### PR TITLE
[ADP-3034] remove BoundedAddressLength class

### PR DESCRIPTION
**Pull request by paolino**
- [x] Add key flavors singletons
- [x] Switch `BoundedAddressLengthClass` to plain function over the singleton
ADP-3034

### CODE REVIEW

**Review by paweljakubas — APPROVED** [review]
lgtm

File: lib/wallet/src/Cardano/Wallet/Primitive/Migration/Selection.hs [hunk 1]
```diff
@@ -280,7 +280,7 @@ assignMinimumAdaQuantity :: TxConstraints -> TokenMap -> TokenBundle
 assignMinimumAdaQuantity constraints m =
     TokenBundle c m
   where
-    -- Using @maxLengthAddressFor $ Proxy @k@ via @constraints@ would not help
+    -- Using @maxLengthAddressFor via @constraints@ would not help
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Primitive/Migration/Selection.hs]:
```suggestion
    -- Using @maxLengthAddressFor@ via @constraints@ would not help
```

**Review by Anviking — APPROVED** [review]
LGTM, just note that a potential downside here is that e.g.:

```haskell
    paymentAddress s@(STestnet _) k = Address
        $ CBOR.toStrictByteString
        $ CBOR.encodeAddress (getKey k)
            [ CBOR.encodeProtocolMagicAttr (magicSNetworkId s)
            ]
```

and

```haskell
    maxLengthAddressFor _ = Address
        $ CBOR.toStrictByteString
        $ CBOR.encodeAddress xpub
            [ CBOR.encodeProtocolMagicAttr (Proto colMagic maxBound)
            ]
      where
        xpub :: CC.XPub
        xpub = CC.toXPub $ CC.generate (BS.replicate 32 0) xprvPass
          where
            xprvPass = mempty :: BS.ByteString
```

now are located further apart, despite being highly related.

File: lib/wallet/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -1617,16 +1619,16 @@ assignChangeAddressesAndUpdateDb ctx argGenChange selection =
         s = getState $ getLatest wallet
         (selectionUpdated, stateUpdated) =
             assignChangeAddresses
-                (defaultChangeAddressGen argGenChange (Proxy @k))
+                (defaultChangeAddressGen argGenChange )
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet.hs:1581]:
```suggestion
                (defaultChangeAddressGen argGenChange)
```

File: lib/wallet/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -1636,7 +1638,7 @@ assignChangeAddressesWithoutDbUpdate ctx argGenChange selection =
         cp <- readCheckpoint
         let (selectionUpdated, _) =
                 assignChangeAddresses
-                    (defaultChangeAddressGen argGenChange (Proxy @k))
+                    (defaultChangeAddressGen argGenChange )
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet.hs]:
```suggestion
                    (defaultChangeAddressGen argGenChange)
```

File: lib/wallet/src/Cardano/Wallet/Address/Keys/BoundedAddressLength.hs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+--
+-- This is useful in situations where we want to compute some function of
+-- an output under construction (such as a minimum UTxO value), but don't
+-- yet have convenient access to a real address.
+--
+-- Please note that this address should:
+--
+--  - never be used for anything besides its length and validity properties.
+--  - never be used as a payment target within a real transaction.
+maxLengthAddressFor
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Address/Keys/BoundedAddressLength.hs:37]: Thank-you for preserving this comment!

File: lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs [hunk 5]
```diff
@@ -2257,7 +2255,7 @@ dummyShelleyChangeAddressGen :: AnyChangeAddressGenWithState
 dummyShelleyChangeAddressGen = AnyChangeAddressGenWithState
     (defaultChangeAddressGen @(SeqState 'Mainnet ShelleyKey )
```
Comment (jonathanknowles) [inline lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs]:
```suggestion
    (defaultChangeAddressGen @(SeqState 'Mainnet ShelleyKey)
```

**Review by jonathanknowles — APPROVED** [review]
Thanks for making this PR!

Just a few suggestions. (See comments.)

File: lib/wallet/src/Cardano/Wallet/Flavor.hs [hunk 6]
```diff
@@ -67,13 +75,41 @@ instance WalletFlavor (RndAnyState n p) where
...
+
+-- | A function to reify the flavor of a key from a state type.
+--
+-- use with
+-- > keyFlavor @s
+keyFlavor :: forall s. WalletFlavor s => KeyFlavorS (KeyOf s)
+keyFlavor = keyOfWallet (walletFlavor @s)
+
+-- | Constraints for a state with a specific key.
+type StateWithKey s k = (WalletFlavor s, KeyOf s ~ k)
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Flavor.hs:115]:
```suggestion
```

File: lib/wallet/src/Cardano/Wallet/Flavor.hs [hunk 7]
```diff
@@ -1,17 +1,23 @@
...
 {-# LANGUAGE TypeFamilies #-}
 
 module Cardano.Wallet.Flavor
     ( WalletFlavorS (..)
     , WalletFlavor (..)
     , KeyOf
     , TestState (..)
+    , KeyFlavorS (..)
+    , keyFlavor
+    , StateWithKey
```
Comment (HeinrichApfelmus) [inline lib/wallet/src/Cardano/Wallet/Flavor.hs:20]:
```suggestion
```

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent, thank you! 😊 Could you remove the leftover `StateWithKey` before merging?

----

The function `maxLengthAddressFor` is "closed world" now — we have an exhaustive list of key flavors, and the function does the right thing for each of them.

For this particular function, I still believe that "open world" would be slightly more appropriate, but I am happy to along with the argument that we should make things closed-world by default, and only open up (via type classes) as needed, instead of the other way around. This is because the "open world by default" tends to make the look look more polymorphic than it actually is, complicating the type signatures while gaining little value from it.

File: lib/wallet/src/Cardano/Wallet/Flavor.hs (hunk 6 above)
Comment (paolino) [reply lib/wallet/src/Cardano/Wallet/Flavor.hs:115]: Will remove on next PR, not to stress CI, thanks for spotting it

**Pull request merged**